File: app/src/main/java/com/example/aaron/recipeassistant/readrecipe/readerservice/RecipeReader.java
package com.example.aaron.recipeassistant.readrecipe.readerservice;

import android.content.Context;
import android.os.Build;
import android.speech.tts.TextToSpeech;
import android.speech.tts.Voice;
import android.util.Log;

import com.example.aaron.recipeassistant.model.Recipe;

import java.util.Locale;
import java.util.Set;

public class RecipeReader {

    //private Context context;
    private TextToSpeech tts;
    private Recipe recipe;

    private int currentDirection = -1;
    private int totalDirections;


    public RecipeReader(Context context) {
        //this.context = context;
        tts = new TextToSpeech(context, new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {
                if (status == TextToSpeech.SUCCESS) {
                    int result = tts.setLanguage(Locale.US);
                    if (result == TextToSpeech.LANG_MISSING_DATA || result == TextToSpeech.LANG_NOT_SUPPORTED) {
                        Log.e("TTS", "This language is not supported");
                    }
                    Set<Voice> voices = tts.getVoices();
                    String voiceName = "en-us-x-sfg#female_2-local";
                    for (Voice tmpVoice : voices) {
                        if (tmpVoice.getName().equals(voiceName)) {
                            tts.setVoice(tmpVoice);
                        }
                    }
                } else {
                    Log.e("TTS", "Initialization Failed!");
                }
            }
        });

    }

    private void speak(final String text, final int queueMode) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            tts.speak(text, queueMode, null, null);
        } else {
            tts.speak(text, queueMode, null);
        }
    }

    public void setRecipe(Recipe recipe) {
        this.recipe = recipe;
        totalDirections = recipe.getDirections().length;
    }

    public void readIngredients() {
        for (String ingredient : recipe.getIngredients()) {
            speak(ingredient, TextToSpeech.QUEUE_ADD);
        }
    }

    private void readDirection() {
        speak(recipe.getDirections()[currentDirection], TextToSpeech.QUEUE_FLUSH);
    }

    public void readFirstDirection() {
        currentDirection = 0;
        readDirection();
    }

    public void readLastDirection() {
        currentDirection = totalDirections - 1;
        readDirection();
    }

    public void readPreviousDirection() {
        currentDirection--;
        if (currentDirection < 0) {
            currentDirection = 0;
        }
        readDirection();
    }

    public void readNextDirection() {
        currentDirection++;
        if (currentDirection >= totalDirections) {
            currentDirection = totalDirections - 1;
        }
        readDirection();
    }

    public void stopReading() {
        tts.stop();
    }

    public void destroy() {
        stopReading();
        tts.shutdown();
    }
}

File: app/src/main/java/com/example/aaron/recipeassistant/readrecipe/readerservice/RecipeReader.kt
package com.example.aaron.recipeassistant.readrecipe.readerservice

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.content.Context
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import java.util.*

class RecipeReader(private val context: Context) : LifecycleObserver {

    private var tts: TextToSpeech? = null
    lateinit var utteranceProgressListener: UtteranceProgressListener

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    private fun initTts() {
        tts = TextToSpeech(context, TextToSpeech.OnInitListener { status ->
            onTtsCreated(status)
        })
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    private fun pause() {
        tts?.stop()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    private fun destroyTts() {
        tts?.shutdown()
        tts = null
    }

    private fun onTtsCreated(status: Int) {
        if (status == TextToSpeech.SUCCESS) {
            val result = tts?.setLanguage(Locale.US)
            if (result == TextToSpeech.LANG_MISSING_DATA || result == TextToSpeech.LANG_NOT_SUPPORTED) {
                Log.e("TTS", "This language is not supported")
            }
            tts?.setOnUtteranceProgressListener(utteranceProgressListener)

        } else {
            Log.e("TTS", "Initialization Failed!")
        }
    }

    fun read(text: String) {
        val params = Bundle()
        params.putString(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, "Speaking")
        tts?.speak(text, TextToSpeech.QUEUE_FLUSH, params, "Speaking")
    }

    fun stopReading() {
        tts?.stop()
    }
}
